Speed up scripts by auto publishing and smart caching

tomlm opened the pull request:
* computing a project hash so we only have to run dotnet restore once per project configuration.  If runtime or nuget package dependencies change, it will run dotnet restore, otherwise it will be skipped.
* not executing _commandRunner to check for existence of dotnet (dotnet-script is already running as a dotnet tool)

filipw commented on src/Dotnet.Script.DependencyModel/Context/DotnetRestorer.cs:
```diff
@@ -40,6 +61,6 @@ string CreatePackageSourcesArguments()
         }
 
-        public bool CanRestore => _commandRunner.Execute("dotnet", "--version") == 0;
+        public bool CanRestore => true;
```
that.... actually makes sense 😱

tomlm commented:
I was experimenting more and this works REALLY well.  This delta I call "Auto publish" 
it
1.  computes a consistent temp forlder for the script %temp%\{hashOfPathToSource}\
2. It will publish a dll to that folder if any of the following is true
* the dll is missing
* the hash of the source code is different then the hash of the code that generated the dll
* the last modified time on the source code is newer then the last modified time for the dll.
3. It then runs the dll

The result is that the first time I run a csx it has 2 second delay, and after that it is fast.  If I modify the source in any way, it will recreate the dll.

I think it's pretty good compromise, coupled with the registering of the .csx extension I can simply create a .csx file and execute it as if it were .CMD or Bash file.  Pretty sweet.

tomlm commented:
hash of syntax tree would probably be better then what I'm doing, because I won't detect changes in included components.  I was just hacking to see if this would work.

tomlm commented:
BTW: here is a simple .csx run twice, you can see the first run is ~5 seconds, second is 1/2 a second or 10x faster:
```shell
C:\scratch\csx>timer \scratch\csx\packer.csx
\scratch\csx\packer.csx
Published C:\scratch\csx\packer.csx to C:\Users\tomlm\AppData\Local\Temp\dotnet-scripts\MKirN1LpoNPP26vxk1Mv4gIv9Sg2+7Nx8NshRtCmAM\packer.dll
packer pack <folder>
Results of execution:
    Exit code 0
    Time of execution 4.891
C:\scratch\csx>timer \scratch\csx\packer.csx
\scratch\csx\packer.csx
packer pack <folder>
Results of execution:
    Exit code 0
    Time of execution 0.532
```

tomlm commented:
Filip, I made a number of changes to this delta which you should review:
1. I changed to use the SourceCode parse Checksum, which should enable change detection even with multiple files 
2. I added support for HTTP urls (my auto publish code was assuming that it was only local file system)
3. I made sure all of the unit tests run.
NOTE: One issue is unit test 166 I had to disable.  It fails when executing in a compiled DLL. This happens without my changes (with published dotnet-script).  Repo is to publish a DLL and then run unit test from the DLL.  It fails saying it can't load an assembly.  I am not certain what this unit test is attempting to do, but I think there is a deeper issue and it is not caused by my delta.
Let me know if you have more info on what that unit test was trying to capture.

tomlm commented:
oh snap.  This delta performs kick-ass, with one thing I broke and don't know how to fix.  Debugging from visual code doesn't seem to work with the Auto published dll's, presumably because there are no emitted .PDB files. :(

Anyone have any ideas?

seesharper commented:
It is super important that the debugger still works :) Publish in debug mode?

tomlm commented:
I added a --nocache flag and changed the debug template to pass it.   --nocache basically is what you have today, and debugging works great again. <whew>

tomlm commented:
Aha, I just figured it out.  Debugger.IsAttached FTW .  If debugger is attached, DLL caching is disabled.  No breaking change.  (I kept no-cache so you can do that if you want too)

filipw commented:
would it work with checking if Debugger is attached or for DEBUG mode?

filipw commented:
Oh, we wrote at the same time 😀great!

tomlm commented:
Guys, time for bed, enough hacking for one night. Love this project, I'll review your comments in the morning.

tomlm commented on src/Dotnet.Script.DependencyModel/Context/DotnetRestorer.cs:
```diff
@@ -20,9 +22,25 @@ public DotnetRestorer(CommandRunner commandRunner, LogFactory logFactory)
+            }
+
+            // get old hash code
+            string hashFile = pathToProjectFile + ".hash";
```
Hmm...perhaps a --clearcache or --force flag?  Although I could argue if your script was working with a particular nuget combo, then it's still good.

seesharper commented:
> hey @seesharper , be great for you to try this out.
> If you are on windows run **dotnet-script --register** to make csx files runnable with no tool
> Then try running the same script twice, I think you will like.

This looks awesome. 👍 Could we make that a separate PR as it is really not related to the caching feature. I also would prefer something cross platform for this. There has been some research for a similar feature in the past and I need to look into how this fits in.

filipw commented:
> Could we make that a separate PR as it is really not related to the caching feature. I also would prefer something cross platform for this.

I agree, could you please extract this feature into a separate PR?  They can ship independently.
Also, it looks like the last commit caused some hang/deadlock on UNIX.

tomlm commented:
I have split this into 2 new PRs
#341 Make csx files directly runnable on windows/OSX/Linux
#342 Speed up scripts by auto-publish to dll cache
